acnportal/algorithms/sorted_algorithms.py

```diff
@@ -21,166 +29,353 @@ class SortedSchedulingAlgo(BaseAlgorithm):
+        continuous_inc=0.1,
+        allow_overcharging=False,
+    ):
+        super().__init__(
+            sort_fn, estimate_max_rate, max_rate_estimator, uninterrupted_charging
+        )
+        self.continuous_inc = continuous_inc
+
+    def round_robin(self, active_sessions, infrastructure):
         """ Schedule EVs using a round robin based equal sharing scheme.
 
         Implements abstract method schedule from BaseAlgorithm.
 
         See class documentation for description of the algorithm.
 
         Args:
-            active_evs (List[EV]): see BaseAlgorithm
+            active_sessions (List[EV]): see BaseAlgorithm
+            infrastructure (InfrastructureInfo): Description of electrical
+                infrastructure.
 
         Returns:
             Dict[str, List[float]]: see BaseAlgorithm
         """
-        continuous_inc = 1
-
-        ev_queue = deque(self._sort_fn(active_evs, self.interface))
-        schedule = {ev.station_id: [0] for ev in active_evs}
-        rate_idx_map = {ev.station_id: 0 for ev in active_evs}
-        allowable_rates = {}
-        for ev in ev_queue:
-            evse_continuous, evse_rates = self.interface.allowable_pilot_signals(
-                ev.station_id
+        queue = deque(self._sort_fn(active_sessions, self.interface))
+        schedule = np.zeros(infrastructure.num_stations)
+        rate_idx = np.zeros(infrastructure.num_stations, dtype=int)
+        allowable_pilots = infrastructure.allowable_pilots.copy()
+        for session in queue:
+            i = infrastructure.get_station_index(session.station_id)
+            # If pilot signal is continuous discretize it with increments of
+            # continuous_inc.
+            if infrastructure.is_continuous[i]:
+                allowable_pilots[i] = np.arange(
+                    session.min_rates[0],
+                    session.max_rates[0] + 1e-7,
+                    self.continuous_inc,
+                )
+            ub = min(
+                session.max_rates[0],
+                infrastructure.max_pilot[i],
+                self.interface.remaining_amp_periods(session),
             )
-            if evse_continuous:
-                evse_rates = np.arange(evse_rates[0], evse_rates[1], continuous_inc)
-            allowable_rates[ev.station_id] = evse_rates
-
-        while len(ev_queue) > 0:
-            ev = ev_queue.popleft()
-            if rate_idx_map[ev.station_id] < len(allowable_rates[ev.station_id]) - 1:
-                schedule[ev.station_id][0] = allowable_rates[ev.station_id][
-                    rate_idx_map[ev.station_id] + 1
-                ]
-                if self.interface.is_feasible(schedule):
-                    rate_idx_map[ev.station_id] += 1
-                    ev_queue.append(ev)
+            lb = max(0, session.min_rates[0])
+            # Remove any charging rates which are not feasible.
+            allowable_pilots[i] = [a for a in allowable_pilots[i] if lb <= a <= ub]
```

Comment:
Since allowable_pilots is already a numpy array, it might be faster to use boolean indexing